src/neighbors.cc

```diff
@@ -1275,32 +1275,35 @@ peerRefreshDNS(void *data)
     eventAddIsh("peerRefreshDNS", peerRefreshDNS, nullptr, 3600.0, 1);
 }
 
-static void
-peerConnectFailedSilent(CachePeer * p)
+// TODO: Move to CachePeer::noteFailure() or similar.
+// TODO: Require callers to detail failures instead of using one (and often
+// misleading!) "TCP" label for all of them.
+void
+peerConnectFailed(CachePeer * const p)
 {
     p->stats.last_connect_failure = squid_curtime;
+    if (p->tcp_up > 0)
+        --p->tcp_up;
 
-    if (!p->tcp_up) {
-        debugs(15, 2, "TCP connection to " << p->host << "/" << p->http_port <<
-               " dead");
-        return;
-    }
+    // TODO: Report peer name. Same-addresses peers often have different names.
 
-    -- p->tcp_up;
+    const auto consideredAliveByAdmin = p->stats.logged_state == PEER_ALIVE;
```

Comment:
When consideredAliveByAdmin is true and tcp_up is zero, Squid itself treats the cache_peer as "dead" or "down" -- the peer selection algorithms will not pick this peer. However, the admin has not been informed about the problem yet. These code changes do not change how Squid tracks/marks dead and "unprobed" peers. They only fix problem _reporting_. There is probably a better way to track peer state, but improving that is outside this (reporting-)bug-fix PR.